crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -1,43 +1,220 @@
+    concurrent_requests_range: RangeInclusive<usize>,
+    /// Maximum amount of received headers to buffer internally.
+    max_buffered_responses: usize,
+    /// The range of headers for body download.
+    header_range: Range<BlockNumber>,
+    /// The last requested block number.
+    last_requested_block_number: Option<BlockNumber>,
+    /// The latest block number returned.
+    latest_queued_block_number: Option<BlockNumber>,
+    /// Requests in progress
+    in_progress_queue: FuturesUnordered<BodiesRequestFuture<B, C>>,
+    /// Buffered responses
+    buffered_responses: BinaryHeap<OrderedBodiesResponse>,
+    /// Queued body responses
+    queued_bodies: Vec<BlockResponse>,
+}
+
+impl<B, C, DB> ConcurrentDownloader<B, C, DB>
+where
+    B: BodiesClient + 'static,
+    C: ConsensusTrait + 'static,
+    DB: Database,
+{
+    fn new(
+        client: Arc<B>,
+        consensus: Arc<C>,
+        db: Arc<DB>,
+        request_batch_size: usize,
+        stream_batch_size: usize,
+        max_buffered_responses: usize,
+        concurrent_requests_range: RangeInclusive<usize>,
+    ) -> Self {
+        Self {
+            client,
+            consensus,
+            db,
+            request_batch_size,
+            stream_batch_size,
+            max_buffered_responses,
+            concurrent_requests_range,
+            header_range: Default::default(),
+            last_requested_block_number: None,
+            latest_queued_block_number: None,
+            in_progress_queue: Default::default(),
+            buffered_responses: Default::default(),
+            queued_bodies: Default::default(),
+        }
+    }
+
+    /// Returns the next contiguous request.
+    fn next_request_fut(&mut self) -> Result<Option<BodiesRequestFuture<B, C>>, db::Error> {
+        let start_at = match self.last_requested_block_number {
+            Some(num) => num + 1,
+            None => self.header_range.start,
+        };
+
+        let request = self.query_headers(start_at, self.request_batch_size)?;
+
+        if request.is_empty() {
+            return Ok(None)
+        }
+
+        self.last_requested_block_number = request.last().map(|h| h.number);
+        Ok(Some(
+            BodiesRequestFuture::new(Arc::clone(&self.client), Arc::clone(&self.consensus))
+                .with_headers(request),
+        ))
+    }
+
+    fn query_headers(
+        &self,
+        start: BlockNumber,
+        count: usize,
+    ) -> Result<Vec<SealedHeader>, db::Error> {
+        let tx = self.db.tx()?;
+
+        // Acquire cursors over canonical and header tables
+        let mut canonical_cursor = tx.cursor_read::<tables::CanonicalHeaders>()?;
+        let mut header_cursor = tx.cursor_read::<tables::Headers>()?;
+
+        // TODO: experimental
+        let mut non_empty_headers = 0;
+        let mut headers = Vec::<SealedHeader>::default();
+        let mut canonical_entry = canonical_cursor.seek_exact(start)?;
```

Comment:
I feel like it would be cleaner if we move all of this to the `HeaderProvider` and replace the DB: Database with : `HeaderProvider`

Reply:
Agreed would be nice but let's do it after we've settled on our utils. Probably will happen naturally as part of adding the RPCs

Reply:
Where would we put `self.stream_batch_size` then? The `HeaderProvider` impls currently are just database wrappers, and don't track any other state.